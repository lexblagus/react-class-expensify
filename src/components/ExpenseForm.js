import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';



//console.log(  moment().format('MMM Do')  );

export default class ExpenseForm extends React.Component{
	constructor(props){
		super(props);
		this.state = {
			id: props.expense ? props.expense.id : undefined,
			description: props.expense ? props.expense.description : '',
			amount: props.expense ? (props.expense.amount/100).toString() : 0,
			notes: props.expense ? props.expense.notes : '',
			createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
			calendarFocused: false,
			error: ''
		};
	}
	onDescChange = (evt) => {
		const description = evt.target.value;
		this.setState(() => ({description}));
	};
	onAmountChange = (evt) => {
		const amount = evt.target.value;
		if(
			!amount ||
			amount.match(/^\d{1,}(\.\d{0,2})?$/)
		){
			//console.log('VALID');
			this.setState(() => ({amount}));
		}else{
			//console.log('INvalid');
		}
	};
	onDateChange = (createdAt) => {
		if( createdAt ){
			this.setState(
				() => ({createdAt})
			);
		}
	};
	onNotesChange = (evt) => {
		const notes = evt.target.value;
		this.setState(() => ({notes}));
	};
	onFocusChange = ({ focused }) => {
		this.setState(
			() => ({ calendarFocused: focused })
		)
	};
	onSubmit = (e) => {
		e.preventDefault();
		
		if(
			!this.state.description ||
			!this.state.amount
		){
			this.setState(
				() => ({ error: 'missing stuff' })
			)
		}else{
			this.setState(
				() => ({ error: '' })
			)
			this.props.onSubmit({
				description: this.state.description,
				amount: parseFloat(this.state.amount, 10) * 100,
				createdAt: this.state.createdAt.valueOf(),
				notes: this.state.notes
			});
		}
	};
	render(){
		return (
			<div>
				{
					this.state.error &&
					<code>{this.state.error}</code>
				}
				<form onSubmit={this.onSubmit}>
					description: 
					<input
						type="text"
						placeholder="description"
						autoFocus
						value={this.state.description}
						onChange={this.onDescChange}
					/>
					<br />
					notes: 
					<textarea
						placeholder="notes"
						value={this.state.notes}
						onChange={this.onNotesChange}
					/>
					<br />
					created at: 
					<SingleDatePicker
						date={this.state.createdAt}
						onDateChange={this.onDateChange}
						focused={this.state.calendarFocused}
						onFocusChange={this.onFocusChange}
						numberOfMonths={1}
						isOutsideRange={()=>false}
					/>
					<br />
					amount: 
					<input
						type="number"
						placeholder="amount"
						value={this.state.amount}
						onChange={this.onAmountChange}
					/>
					<br />
					<button>Save</button>
				</form>
			</div>
		);
	}
	
}