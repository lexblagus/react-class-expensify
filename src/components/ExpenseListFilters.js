import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByDate, sortByAmount } from '../actions/filters';


const ExpenseListFilters = (props) => (
	<div>
		<input
			type="text"
			value={props.filters.text}
			onChange={(e) => {
				props.dispatch(setTextFilter(e.target.value));
			}}
			
		/>
		<select
			value={props.filters.sortBy}
			onChange={(evt) => {
				if(evt.target.value === 'date'){
					props.dispatch(sortByDate());
				} else if(evt.target.value === 'amount'){
					props.dispatch(sortByAmount());
				}
			}}
		>
			<option value="amount">Amount</option>
			<option value="date">Date</option>
		</select>
	</div>
);

const mapStateToProps = (state) => ({
	filters: state.filters
});

export default connect(mapStateToProps)(ExpenseListFilters);
