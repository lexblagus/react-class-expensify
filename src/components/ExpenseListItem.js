import React from 'react';
import { connect } from 'react-redux';
import { removeExpense } from '../actions/expenses';


const ExpenseListItem = ({dispatch, id, description, amount, createdAt}) => (
	<div>
		<h2>{description}</h2>
		<button
			onClick = {() => {
				dispatch(removeExpense({id}));
			}}
		>remove</button>
		&nbsp;
		amount = {amount},
		createdAt = {createdAt}
	</div>
);

export default connect()(ExpenseListItem);
