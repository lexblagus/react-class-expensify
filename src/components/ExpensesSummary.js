import React from 'react';
import expensesTotal from '../selectors/expenses-total';
import selectExpenses from '../selectors/expenses';
import { connect } from 'react-redux';

const ExpensesSummary = (props) => (
	<div>
		<h3 className="title is-3">
			List
			<span className="title is-4">
				{' '}
				with
				{' '}
				{props.qtt}
				{' '}
				item{props.qtt>1?'s':''}
				{' '}
				$ {props.tot}
			</span>
		</h3>
	</div>
);

const mapStateToProps = (state) => ({
		'tot': expensesTotal(selectExpenses(state.expenses,state.filters)),
		'qtt': selectExpenses(state.expenses,state.filters).length
});

//export default ExpensesSummary;
export default connect(mapStateToProps)(ExpensesSummary);
