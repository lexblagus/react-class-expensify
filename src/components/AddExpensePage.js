import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';


const AddExpensePage = (props) => (
	<div>
		<div className="hero is-light">
			<h1 className="title is-1">Add expense</h1>
			<ExpenseForm
				onSubmit = {
					(expense)=>{
						props.dispatch(
							addExpense(expense)
						);
						props.history.push('/');
					}
				}
			/>
		</div>
	</div>
);

export default connect()(AddExpensePage);
