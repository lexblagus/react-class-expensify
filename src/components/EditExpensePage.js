import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

const EditExpensePage = (props) => {
	return (
		<div>
			<section className="section">
				<div className="container">
					{
						props.expense &&
						<div>
							<h2 className="title is-2">
								Edit
								<span className="title is-4">
									&nbsp;
									{props.expense.id}
								</span>
							</h2>
							<div className="card">
								<div className="card-content">
									<ExpenseForm
										expense={props.expense}
										onSubmit={
											(expense)=>{
												console.log('updated expense',expense);
												props.dispatch(
													editExpense(
														props.expense.id,
														{...expense}
													)
												);
												props.history.push('/');
											}
										}
									/>
									<button
										className="button is-small is-danger"
										onClick = {() => {
											props.dispatch(
												removeExpense(
													{id: props.expense.id}
												)
											);
											props.history.push('/');
										}}
									>Delete</button>
								</div>
							</div>
						</div>
					}
					{
						!props.expense &&
						<div>
							<h2 className="title is-2">
								Item not found
							</h2>
						</div>
					}
				</div>
			</section>
		</div>
	);
}

const mapStateToProps = (state, props) => ({
	expense: state.expenses.find(
		(expense) => (expense.id === props.match.params.id)
	)
});

export default connect( mapStateToProps )( EditExpensePage );
