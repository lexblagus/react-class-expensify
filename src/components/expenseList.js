import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
	<div className="container">
		<h3 className="title is-3">
			List
			<span className="title is-4">
				{' '}
				with
				{' '}
				{props.expenses.length}
				{' '}
				item{props.expenses.length>1?'s':''}
			</span>
		</h3>
		{
			props.expenses.length === 0 ? (
				<div className="notification is-warning">
					nothing right now
				</div>
			) : (
				props.expenses.map(
					(expense) => (
						<ExpenseListItem key={expense.id} {...expense} />
					)
				)
			)
		}
	</div>
);

const mapStateToProps = (state) => ({
	expenses: selectExpenses(state.expenses,state.filters)
});


export default connect(mapStateToProps)(ExpenseList);
