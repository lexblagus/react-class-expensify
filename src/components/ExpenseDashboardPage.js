import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from './ExpensesSummary';


const ExpenseDashboardPage = () => (
	<div>
		<section className="section">
			<div className="container">
				<h2 className="title is-2">View</h2>
				<div className="card">
					<div className="card-content">
						<ExpenseListFilters />
						<ExpensesSummary />
						<ExpenseList />
					</div>
				</div>
			</div>
		</section>
	</div>
);

export default ExpenseDashboardPage;
