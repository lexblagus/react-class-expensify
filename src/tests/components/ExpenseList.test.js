import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseList }  from '../../components/expenseList';
import dummyStore01 from '../fixtures/dummyStore01';


test('Should render ExpenseList with expenses', ()=>{
	const wrapper = shallow(
		<ExpenseList expenses={dummyStore01} />
	);
	
	expect(wrapper).toMatchSnapshot();
});

test('Should render empty ExpenseList', ()=>{
	const wrapper = shallow(
		<ExpenseList expenses={[]} />
	);
	
	expect(wrapper).toMatchSnapshot();
});