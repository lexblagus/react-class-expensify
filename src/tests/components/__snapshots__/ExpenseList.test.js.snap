// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpenseList with expenses 1`] = `
<div
  className="container"
>
  <ExpenseListItem
    amount={10000}
    createdAt={1252015000}
    description="1 water bill"
    id="2da95af6-3181-4902-a9fd-ec52eff5203c"
    key="2da95af6-3181-4902-a9fd-ec52eff5203c"
    notes=""
  />
  <ExpenseListItem
    amount={109500}
    createdAt={1338415000}
    description="2 gas bill"
    id="c9b3ec51-3b1b-40a3-904f-76ef27f79c80"
    key="c9b3ec51-3b1b-40a3-904f-76ef27f79c80"
    notes="gas gas gas gas gas"
  />
  <ExpenseListItem
    amount={0}
    createdAt={1424815000}
    description="3 rent"
    id="e53d429d-0312-4dd3-aa8c-ed55d2229ae6"
    key="e53d429d-0312-4dd3-aa8c-ed55d2229ae6"
    notes="rent rent rent rent "
  />
  <ExpenseListItem
    amount={100}
    createdAt={1511215000}
    description="Description A"
    id="5d94f325-5387-4406-9c42-45f250350180"
    key="5d94f325-5387-4406-9c42-45f250350180"
    notes="notes notes notes 1 "
  />
  <ExpenseListItem
    amount={200}
    createdAt={1597615000}
    description="Description B"
    id="adb329fa-d15a-4605-8e09-44f544a69230"
    key="adb329fa-d15a-4605-8e09-44f544a69230"
    notes="notes notes notes 2"
  />
  <ExpenseListItem
    amount={300}
    createdAt={1684015000}
    description="Description C"
    id="09afc748-8363-4a58-a16c-62bf952622fb"
    key="09afc748-8363-4a58-a16c-62bf952622fb"
    notes="notes notes notes 3"
  />
  <ExpenseListItem
    amount={400}
    createdAt={1770415000}
    description="Description D"
    id="cea054a4-243d-42d3-a11c-dacd957df150"
    key="cea054a4-243d-42d3-a11c-dacd957df150"
    notes="notes notes notes 4"
  />
  <ExpenseListItem
    amount={500}
    createdAt={1856815000}
    description="Description E"
    id="1d42f8e6-0cc4-402f-a5cf-6ebf33d845f9"
    key="1d42f8e6-0cc4-402f-a5cf-6ebf33d845f9"
    notes="notes notes notes 5"
  />
</div>
`;

exports[`Should render empty ExpenseList 1`] = `
<div
  className="container"
>
  <div
    className="notification is-warning"
  >
    nothing right now
  </div>
</div>
`;
